### PR TITLE
Nnet3 dropout: code for Fast LSTM and scripts for AMI etc

### PR DESCRIPTION
**Pull request by GaofengCheng**
This PR will include：
1. code for  fast lstm dropout (based on dpovey's PR1387)
2. scripts with dropout for AMI-IHM, AMI-SDM, SWBD and TEdlium (tdnn-(fast)lstms and b(fast)lstms)

Time schedule:
1. fast lstm dropout code
2. scripts for  
                     AMI-IHM-> (done)
                     AMI-SDM-> (done)
                     SWBD->  (done)
                     TEdlium-> (done)
                     (follow the time axis)

### CODE REVIEW

File: egs/ami/s5b/RESULTS_ihm [hunk 1]
```diff
@@ -84,6 +84,20 @@
 %WER 20.8 | 13098 94489 | 82.0 10.0 8.0 2.8 20.8 53.2 | -0.096 | exp/ihm/chain_cleaned/tdnn_lstm1i_sp_bi_ld5/decode_dev/ascore_11/dev_hires.ctm.filt.sys
 %WER 20.7 | 12643 89980 | 81.7 11.5 6.8 2.5 20.7 51.8 | 0.015 | exp/ihm/chain_cleaned/tdnn_lstm1i_sp_bi_ld5/decode_eval/ascore_11/eval_hires.ctm.filt.sys
 
+# local/chain/tuning/run_tdnn_lstm_1l.sh --mic ihm  --train-set train_cleaned  --gmm tri3_cleaned
```
Comment (danpovey) [inline egs/ami/s5b/RESULTS_ihm]: At this level, please just include the results for the "recommended" system which is 1m.
You should put all the comparative results in the individual scripts inside local/chain/tuning.
Use the standard compare_wer.sh script, whatever it's called, and also include the output
of chain_dir_info.pl from each of those scripts, in a comment in that script.

Comment (GaofengCheng) [inline egs/ami/s5b/RESULTS_ihm]: @danpovey OK, will do

File: egs/ami/s5b/local/chain/tuning/run_tdnn_lstm_1l.sh [hunk 2]
```diff
@@ -5,11 +5,14 @@
 #System               tdnn_lstm1i_sp_bi_ld5 tdnn_lstm1l_sp_bi_ld5
 #WER on dev        20.6      19.8
 #WER on eval        20.1      19.2
-#Final train prob      -0.045 -0.067
```
Comment (danpovey) [inline egs/ami/s5b/local/chain/tuning/run_tdnn_lstm_1l.sh]: please give a bit more context in all of these scripts, at the top e.g. in this case where it says
` # same as 1i but with per-frame dropout on LSTM layer`
about what this is, e.g....
` # This (1l.sh) is thesame as 1i but with per-frame dropout on LSTM layer
  # It is a regular (not-fast) LSTM with per-frame dropout on [which gates?].`
And explain how it relates to the paper, e.g. is it "place4" in the paper?
You can send me the final pdf to put on my publications page if it's not already there,
and include a link from the script to the paper.

Comment (GaofengCheng) [inline egs/ami/s5b/local/chain/tuning/run_tdnn_lstm_1l.sh]: @danpovey I have emailed the final pdf to you, later will go on updating this PR

File: egs/ami/s5b/RESULTS_ihm [hunk 3]
```diff
@@ -84,6 +84,10 @@
 %WER 20.8 | 13098 94489 | 82.0 10.0 8.0 2.8 20.8 53.2 | -0.096 | exp/ihm/chain_cleaned/tdnn_lstm1i_sp_bi_ld5/decode_dev/ascore_11/dev_hires.ctm.filt.sys
 %WER 20.7 | 12643 89980 | 81.7 11.5 6.8 2.5 20.7 51.8 | 0.015 | exp/ihm/chain_cleaned/tdnn_lstm1i_sp_bi_ld5/decode_eval/ascore_11/eval_hires.ctm.filt.sys
 
+# local/chain/tuning/run_tdnn_lstm_1m.sh --mic ihm  --train-set train_cleaned  --gmm tri3_cleaned
```
Comment (vijayaditya) [inline egs/ami/s5b/RESULTS_ihm]: Does 1i already account for all the new changes like fast-lstm component. If not please specify what all things changed between 1i and 1m.

File: egs/ami/s5b/local/chain/tuning/run_tdnn_lstm_1l.sh [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+#IHM
+#System               tdnn_lstm1i_sp_bi_ld5 tdnn_lstm1l_sp_bi_ld5
+#WER on dev        20.6      19.8
+#WER on eval        20.1      19.2
+#Final train prob      -0.044763 -0.0666221
+#Final valid prob     -0.0981107 -0.097616
+#Final train prob (xent)     -0.722765 -0.915559
+#Final valid prob (xent)      -1.03985  -1.09907
+
+# steps/info/chain_dir_info.pl exp/ihm/chain_cleaned/tdnn_lstm1i_sp_bi_ld5/ exp/ihm/chain_cleaned/tdnn_lstm1l_sp_bi_ld5/
```
Comment (vijayaditya) [inline egs/ami/s5b/local/chain/tuning/run_tdnn_lstm_1l.sh]: Please specify the set of flags to be used to recreate these results.

File: egs/ami/s5b/local/chain/tuning/run_tdnn_lstm_1m.sh [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+#IHM
+#System            tdnn_lstm1j_sp_bi_ld5 tdnn_lstm1m_sp_bi_ld5
+#WER on dev        20.8      19.9
+#WER on eval        20.3      19.3
+#Final train prob     -0.0439145 -0.0653269
+#Final valid prob       -0.10673 -0.0998743
+#Final train prob (xent)     -0.683776 -0.884698
+#Final valid prob (xent)      -1.05254  -1.09002
+
+# steps/info/chain_dir_info.pl exp/ihm/chain_cleaned/tdnn_lstm1j_sp_bi_ld5/ exp/ihm/chain_cleaned/tdnn_lstm1m_sp_bi_ld5/
```
Comment (vijayaditya) [inline egs/ami/s5b/local/chain/tuning/run_tdnn_lstm_1m.sh]: Please specify the flags to be specified to recreate these results.

File: egs/wsj/s5/steps/libs/nnet3/xconfig/lstm.py [hunk 6]
```diff
@@ -833,14 +831,16 @@ def generate_lstm_config(self):
         configs.append("# The core LSTM nonlinearity, implemented as a single component.")
         configs.append("# Input = (i_part, f_part, c_part, o_part, c_{t-1}), output = (c_t, m_t)")
         configs.append("# See cu-math.h:ComputeLstmNonlinearity() for details.")
-        configs.append("component name={0}.lstm_nonlin type=LstmNonlinearityComponent cell-dim={1} {2}".format(name, cell_dim, lstm_str))
+        configs.append("component name={0}.lstm_nonlin type=LstmNonlinearityComponent cell-dim={1} "
```
Comment (vijayaditya) [inline egs/wsj/s5/steps/libs/nnet3/xconfig/lstm.py]: using named arguments is preferable when you have so many fields.

File: src/cudamatrix/cu-kernels.cu [hunk 7]
```diff
@@ -3042,17 +3053,27 @@ static void _diff_lstm_nonlinearity(const int cell_dim, const int num_rows,
...
       const Real i_part = input[i * input_stride + j];
       const Real f_part = input[i * input_stride + j + cell_dim];
       const Real c_part = input[i * input_stride + j + 2 * cell_dim];
       const Real o_part = input[i * input_stride + j + 3 * cell_dim];
       const Real c_prev = input[i * input_stride + j + 4 * cell_dim];
 
-      const Real i_t = 1 / (1 + exp(-i_part - w_ic * c_prev));
-      const Real f_t = 1 / (1 + exp(-f_part - w_fc * c_prev));
+
+      const Real i_scale = (have_dropout_mask ?
```
Comment (vijayaditya) [inline src/cudamatrix/cu-kernels.cu]: NIT: `has_dropout_mask` might be easier to follow than `have_dropout_mask`.

File: egs/wsj/s5/steps/libs/nnet3/xconfig/lstm.py (hunk 6 above)
Comment (GaofengCheng) [inline egs/wsj/s5/steps/libs/nnet3/xconfig/lstm.py]: @vijayaditya Sorry Vijay, I'm not quite clear how to use named arguments here, can you give some detailed examples?

**Comment by danpovey** [comment]
What he means is something like
```
 "component name={name}.lstm_nonlin... ".format(name=name, ...)
```

**Comment by GaofengCheng** [comment]
@danpovey PR for SWBD and Tedlium will take about 2~3 days

**Comment by danpovey** [comment]
@GaofengCheng, how much longer till this is ready to merge?

**Pull request merged**